oak_functions/location_utils/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Utilities for finding and comparing longitude- and latitude-based locations.
+
+#[cfg(test)]
+mod tests;
+
+const EARTH_RADIUS: f32 = 6_371_000.0;
+
+/// Represents a cell on a sphere.
+#[derive(Debug, PartialEq)]
+pub struct Cell {
```

Comment:
If we are going to have a custom cell representation, have you (re)considered S2? e.g. https://docs.rs/s2/0.0.10/s2/cellid/struct.CellID.html . Even if you have excluded it, could you add a comment somewhere to mention it and the reason why it was excluded? I think it would be useful for future maintainers so that they don't end up re-evaluating it just to find out why it would not work, for instance.

Reply:
I have considered it twice before and chose this custom cell definition for two main reasons:

  - Attempt to minimise the overall number of calculations (and specifically floating point calculations) in the Wasm code for performance reasons
  - Size and complexity of the S2 geometry libraries (we only need a tiny fraction of the functionality)

But having looked at it once again, it might be possible to move much of the calculation into the lookup data generation process and keep the number Wasm calculations relatively low. Presumably the compiler should also strip out the code for unused features from the library, so the size of the compiled Wasm might not be that much bigger.

I will do a separate PR using S2 cells and do some benchmarking to see how much difference in Wasm size and speed there might be, and then we can decide on a way forward.

Reply:
Created #2296 to track this.